Microchip: MEC172x: pinctrl driver

jvasanth1 opened the pull request:
Implemented pinctrl driver for MEC172x device

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

PR moved to #43282

gmarull reviewed (changes requested):
Thanks for your proposal. There are too many changes in a single commit, please split PR into smaller commits to ease review. For example: new driver, board porting, old driver removal.

jvasanth1 commented:
Hello @gmarull 
Could you elaborate on splitting into multiple commits? We have seen before that multiple commits getting a push back.

Is this breakdown (of 4 commits) okay?
1. pinctrl driver (along with soc folder changes)
2. MEC172x board changes to support pinctrl driver
3. Update existing drivers to use pinctrl (adc, espi, gpio, i2c, serial, qmspi)
4. Remove pinmux driver

gmarull commented on drivers/adc/adc_mchp_xec_v2.c:
```diff
@@ -304,45 +307,66 @@ static int adc_xec_init(const struct device *dev)
 		| XEC_ADC_CTRL_SINGLE_DONE_STATUS
 		| XEC_ADC_CTRL_REPEAT_DONE_STATUS;
 
+	cfg->irq_config_func();
+
```
an example: this change has nothing to do with pinctrl

gmarull commented on drivers/adc/adc_mchp_xec_v2.c:
```diff
@@ -304,45 +307,66 @@ static int adc_xec_init(const struct device *dev)
-		    &adc_xec_dev_data_0, &adc_xec_dev_cfg_0,
-		    PRE_KERNEL_1, CONFIG_ADC_INIT_PRIORITY,
-		    &adc_xec_api);
+#define ADC_XEC_DEVICE(i)						\
```
another example: converting drivers to be multi-instance is ok, but this needs to have its own commit

gmarull commented on drivers/gpio/gpio_mchp_xec_v2.c:
```diff
@@ -125,10 +135,12 @@ static int gpio_xec_configure(const struct device *dev,
 	mask |= MCHP_GPIO_CTRL_PUD_MASK;
 
+	/* if both pulls are enabled the pin is in repeater mode */
 	if ((flags & GPIO_PULL_UP) != 0U) {
 		/* Enable the pull and select the pullup resistor. */
 		pcr1 |= MCHP_GPIO_CTRL_PUD_PU;
-	} else if ((flags & GPIO_PULL_DOWN) != 0U) {
+	}
+	if ((flags & GPIO_PULL_DOWN) != 0U) {
```
this is changing GPIO driver behavior: also deserves a commit explaining why

jvasanth1 replied:
I am planning to focus only on pinctrl changes for this commit. So will remove changes not specific to pinctrl. 
Do we still need multiple commits, please let me know?

gmarull replied:
Yes, you can take as a reference implementations already merged or in progress listed in https://github.com/zephyrproject-rtos/zephyr/issues/39740

jvasanth1 replied:
Thanks for the examples; will update this pull request with multiple commits for easy review.

jvasanth1 replied:
I have added the pinctrl changes as a series of 9 commits.

jvasanth1 committed:
drivers: pinctrl: Microchip MEC172x pinctrl driver

Add core files for Microchip MEC172x pinctrl driver

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
dts: Update MEC172x pinctrl dts

Update dtsi and dts bindings for pinctrl driver

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
gpio: MEC172x: update gpio module for pinctrl

Changes to gpio module to support pinctrl

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
i2c: update MEC172x i2c module for pinctrl

Changes to i2c module to support pinctrl

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
espi: updated MEC172x espi module for pinctrl

Changes to espi module to support pinctrl

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
serial: update mchp uart module to use pinctrl

Changes to uart module to support pinctrl

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
qmspi: update MEC172x qmspi module for pinctrl

Changes to qmspi module to support pinctrl

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
pinmux: Remove pinmux.c source file

Since we are moving to pinctrl, removing pinmux.c
from mec172x board folder and removing pinmux from dts

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

gmarull commented on dts/arm/microchip/mec172x/mec172xnsz-pinctrl.dtsi:
```diff
@@ -0,0 +1,966 @@
+
+#include <dt-bindings/pinctrl/mchp-xec-pinctrl.h>
+
+/ {
+	soc {
+		pinctrl: pin-controller@40081000 {
```
nit: you can just do `&pinctrl {`

gmarull commented on dts/arm/microchip/mec172x/mec172xnsz-pinctrl.dtsi:
```diff
@@ -0,0 +1,966 @@
+	soc {
+		pinctrl: pin-controller@40081000 {
+
+			/* ADC */
+			pinctrl_adc0_in0: adc0_in0 {
+				pinmux = < MCHP_XEC_PINMUX(0200, AF1) >;
+			};
+
+			pinctrl_adc0_in1: adc0_in1 {
+				pinmux = < MCHP_XEC_PINMUX(0201, AF1) >;
+			};
+
+			pinctrl_adc0_in2: adc0_in2 {
+				pinmux = < MCHP_XEC_PINMUX(0202, AF1) >;
+			};
+
```
note, non-blocking: consider using the grouping approach (is less verbose when hand-written). For example, you'd have:

```
&adc0_default {
    group1 {
        pinmux = <MCHP_XEC_PINMUX(0200, AF1)>,
                        <MCHP_XEC_PINMUX(0201, AF1)>,
                        <MCHP_XEC_PINMUX(0202, AF1)>,
                        <MCHP_XEC_PINMUX(0203, AF1)>,
                        <MCHP_XEC_PINMUX(0204, AF1)>,
                        <MCHP_XEC_PINMUX(0205, AF1)>,
                        <MCHP_XEC_PINMUX(0206, AF1)>,
                        <MCHP_XEC_PINMUX(0207, AF1)>,
                        <MCHP_XEC_PINMUX(067, AF1)>;
    };
};
```

and then

```
&adc0 {
    pinctrl-0 = <&adc0_default>;
    pinctrl-names = "default";
};
```

The current pinctrl file is ~1K lines, which seems too much really. The approach used here works better when autogeneration is done.

gmarull replied:
another comment: please use a better name for nodes if going with current approach, e.g

```
adc0_in0_default
```

^- peripheral adc0, signal in0, default state

jvasanth1 replied:
We don't want to group the pins since we want to allow selective configuration of the pins in the supposed group. For example for adc, we will not be enabling all the adc channels. Most customers are not enabling all the adc channels all at once.

Also, other pinctrl implementations (of other vendors) are not using group feature.

gmarull replied:
> We don't want to group the pins since we want to allow selective configuration of the pins in the supposed group. For example for adc, we will not be enabling all the adc channels. Most customers are not enabling all the adc channels all at once.

The same functionality can be achieved with groups, it's just about representation formats.

jvasanth1 replied:
Not considering grouping approach for now.

gmarull commented on dts/bindings/pinctrl/microchip,xec-pinctrl-conf.yaml:
```diff
@@ -0,0 +1,48 @@
+    Microchip XEC Pin controller Node
+    Based on pincfg-node.yaml binding.
+
+    Note: `bias-disable` and `drive-push-pull` are default pin configurations.
+           They will be applied in case no `bias-foo` or `driver-bar` properties
+           are set.
```
please try to extend this documentation as other bindings do, refer to nrf/gd32/ etc.

jvasanth1 replied:
okay

jvasanth1 replied:
updated documentation

gmarull commented on dts/bindings/pinctrl/microchip,xec-pinctrl-conf.yaml:
```diff
@@ -0,0 +1,48 @@
+          type: string
+          default: "low-speed"
+          enum:
+          - "low-speed"           # Default value.
```
comment is redundant, description provides info.

jvasanth1 replied:
will remove comment

gmarull commented on dts/bindings/pinctrl/microchip,xec-pinctrl-conf.yaml:
```diff
@@ -0,0 +1,48 @@
+          enum:
+          - "low-speed"           # Default value.
+          - "high-speed"
+          description: Pin speed. Default to low-speed.
```
please justify default here: is it SoC default?

gmarull commented on dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:
```diff
@@ -0,0 +1,63 @@
+          type: string
+          default: "hw-default"
+          enum:
+          - "hw-default"        # Default value
+          - "low-speed"
+          - "high-speed"
+          description: Pin speed.
```
what does "hw-default" mean?

jvasanth1 replied:
hw-default refers to the hardware default slew-rate config; for the device MEC172x, this hardware default is configurable through OTP option, hence we refer to that default here.

gmarull replied:
Please, describe this hardware feature in detail, it's not clear from the bindings docs that these SoCs have this capability.

gmarull replied:
![image](https://user-images.githubusercontent.com/25011557/152317113-b4df2541-6d37-491f-90d6-ead7207c02c8.png)
 seems like an edge case that should be documented here, too.

jvasanth1 replied:
will update the yaml doc to include following:

```
Some gpio pins do not implement all features. For example, the eSPI 
pins do not support drive strength or slew-rate configurability.
Please check datasheet to confirm. 
```

gmarull commented on dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:
```diff
@@ -0,0 +1,63 @@
+        drive-strength:
+          required: false
+          type: string
+          default: "hw-default"
+          enum:
+            - "hw-default"      # Default value
```
ditto

gmarull replied:
I see 2 valid sets, can you please clarify:

![image](https://user-images.githubusercontent.com/25011557/152316332-15d1fb6e-152e-41c7-b7a9-6536af4636a6.png)

jvasanth1 replied:
Table 2-2 indicates the PIO (buffer) type for the pins. The first set of values is for pins with buffer type PIO, the second set of values applies to pins with buffer type PIO-24.

Will update doc to include this information.

gmarull commented on drivers/gpio/gpio_mchp_xec_v2.c:
```diff
@@ -368,6 +378,110 @@ static void gpio_gpio_xec_port_isr(const struct device *dev)
+ * These are used by the PINCTRL driver and other users.
+ */
+
+/*
+ * Called by PINCTRL
+ * dev = pointer to GPIO port device
+ * pin = pin bit position in GPIO port
+ * cfg = PINCTRL encoded pin properties:
+ *       Pulls, output buffer type, slew rate, drive strength
+ * altf = alternate function where 0 = GPIO
+ */
+
+static void config_drive_slew(const struct device *dev, int pin, int conf)
```
please use proper docstrings format (Doxygen)

jvasanth1 replied:
will update

gmarull commented on drivers/i2c/i2c_mchp_xec_v2.c:
```diff
@@ -102,6 +103,10 @@ struct i2c_xec_data {
 	uint8_t i2c_status;
 };
 
+#define DRV_CONFIG(dev) ((const struct i2c_xec_config *)(dev)->config)
+#define DRV_DATA(dev) ((struct i2c_xec_data *)(dev)->data)
+#define HAL_INSTANCE(dev) (struct i2c_smb_regs *)(DRV_CONFIG(dev)->base_addr)
```
please no more DRV_CONFIG/DRV_DATA, tree is now finally free of all that crap

jvasanth1 replied:
okay, will remove

gmarull commented on soc/arm/microchip_mec/common/soc_i2c.c:
```diff
@@ -0,0 +1,108 @@
+ * We GPIO control register always active RO pad input bit.
+ * lines b[0]=SCL pin state at pad, b[1]=SDA pin state at pad
+ */
+int soc_i2c_port_lines_get(uint8_t port, uint32_t *lines)
```
why is i2c specific stuff getting inserted in `soc/`? shouldn't that be in I2C driver?

jvasanth1 replied:
mchp i2c driver needs to read the pin states; the correct pins is dependent on the port. We don't want to manage the port-gpio mappings (which is soc specific) in the driver, hence these functions are in soc/

gmarull commented on drivers/adc/adc_mchp_xec_v2.c:
```diff
@@ -304,45 +307,66 @@ static int adc_xec_init(const struct device *dev)
 		| XEC_ADC_CTRL_SINGLE_DONE_STATUS
 		| XEC_ADC_CTRL_REPEAT_DONE_STATUS;
 
+	cfg->irq_config_func();
```
please split this change into 2 commits: one that makes driver multi-instance and 2nd pinctrl. If not possible for some reason please explain changes in more detail in the commit message

jvasanth1 replied:
ok, will split this commit into 2 commits

gmarull reviewed (changes requested):
Thanks for splitting up changes, I have a few comments please check.

gmarull commented on boards/arm/mec172xevb_assy6906/mec172xevb_assy6906_defconfig:
```diff
@@ -12,6 +12,7 @@ CONFIG_RTOS_TIMER=y
 CONFIG_CLOCK_CONTROL=y
 CONFIG_GPIO=y
 CONFIG_PINMUX=y
+CONFIG_PINCTRL=y
```
Btw, `CONFIG_PINMUX` should be removed

jvasanth1 replied:
> if all boards need, in practice, `CONFIG_PINCTRL=y`, better enable it by default at SoC level, see for example GD32

This is our first pinctrl driver, we want to focus on mec172xevb_assy6906 board before applying pinctrl on other boards.

jvasanth1 replied:
> Btw, `CONFIG_PINMUX` should be removed

will remove

jvasanth1 replied:
CONFIG_PINMUX removed in the final commit of this PR

jvasanth1 commented on drivers/gpio/gpio_mchp_xec.h:
```diff
@@ -0,0 +1,33 @@
+ * @file header for Microchip XEC GPIO
+ */
+
+#include <drivers/gpio.h>
```
yes, will update

jvasanth1 commented on drivers/gpio/gpio_mchp_xec.h:
```diff
@@ -0,0 +1,33 @@
+
+/* GPIO buses definitions */
+
+#define MCHP_XEC_GPIO_PORT(pin) (((pin) >> 5) & 0x7)
```
okay

jvasanth1 commented on drivers/pinctrl/pinctrl_mchp_xec.c:
```diff
@@ -0,0 +1,83 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/pinctrl.h>
+#include <gpio/gpio_mchp_xec.h>
+#include <soc.h>
```
will add

jvasanth1 commented on soc/arm/microchip_mec/common/soc_gpio.h:
```diff
@@ -20,4 +20,9 @@
 #define MCHP_GPIO_MAX_PORT    6
 #define MCHP_NUM_GPIO_PORTS   6
 
+struct soc_gpio_pinctrl {
+	uint32_t pinmux;
+	uint32_t pincfg;
+};
```
this is not redundant, will be removed

jvasanth1 commented on soc/arm/microchip_mec/CMakeLists.txt:
```diff
@@ -1,3 +1,8 @@
 add_subdirectory(${SOC_SERIES})
+add_subdirectory(common)
+
+# if (DEFINED SOC_SERIES_MEC172X)
+#zephyr_include_directories(common)
+#endif()
```
we are adding some header files in common directory which could be used by future devices as well. Also the common folder includes soc_i2c.c (added as a separate folder) which needs to be compiled.

tejlmand replied:
I think @gmarull questions why there is code that is commented out.

Please remove, such code can be added later when there is a need.

jvasanth1 commented on drivers/pinctrl/pinctrl_mchp_xec.c:
```diff
@@ -0,0 +1,83 @@
+
+		func = MCHP_XEC_PINMUX_FUNC(mux);
+		if (func >= AFMAX) {
+			__ASSERT_NO_MSG(func);
```
will change to return -EINVAL

jvasanth1 commented on soc/arm/microchip_mec/common/soc_dt.h:
```diff
@@ -0,0 +1,25 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _MICROCHIP_XEC_SOC_DT_H_
```
This contains the pin configuration values used by the utility macros in pinctrl_soc.h

jvasanth1 committed:
qmspi: MEC172x: fix unused variable warning

Fix unused variable warning when CONFIG_SPI_ASYNC
is not defined

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
adc: updates to MEC172x adc to support pinctrl

Changes to adc module to support pinctrl.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 commented:
Updated to address most of the review comments. Please check.

jvasanth1 commented on drivers/pinctrl/pinctrl_mchp_xec.c:
```diff
@@ -0,0 +1,83 @@
+ *
+ * Entries will be NULL if the GPIO port is not enabled.
+ */
+static const struct device * const gpio_ports[] = {
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio_000_036)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio_040_076)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio_100_136)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio_140_176)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio_200_236)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio_240_276)),
+};
```
Can you suggest some example of phandle that I could refer? 

Also is it possible to defer the suggested change once our basic pinctrl changes are merged into zephyr main. We could take up the suggested change as a separate PR later.

jvasanth1 replied:
Could we move this table to the chip folder in the soc?
If so, do we need a soc API to get the table pointer and size?

jvasanth1 replied:
Added phandle in device tee.

jvasanth1 replied:
Reverted to gpio_ports[] table approach as per the discussion in the [comment](https://github.com/zephyrproject-rtos/zephyr/pull/41888#discussion_r809787897).

jvasanth1 commented on dts/arm/microchip/mec172x/mec172xnsz-pinctrl.dtsi:
```diff
@@ -0,0 +1,794 @@
+		pinctrl: pin-controller@40081000 {
+
+			/* ADC */
+			pinctrl_adc0_in0: adc0_in0 {
+				pinmux = < MCHP_XEC_PINMUX(0200, MCHP_AF1) >;
+			};
```
Our datasheet is based on 3-digit GPIO octal numbering for the pins. Currently, we have defined the pins based on the different alternate functions as it applies to the peripheral. 

Can we use the following naming convention (remove pinctrl, add gpio to match the datasheet convention) ?

```
gpio200_adc0_in0
gpio201_adc0_in1
```

gmarull replied:
You should also consider how you are going to support multiple states, for example the “sleep” state, pre-defined nodes can only be used for a single state.

gmarull replied:
> Our datasheet is based on 3-digit GPIO octal numbering for the pins. Currently, we have defined the pins based on the different alternate functions as it applies to the peripheral.
> 
> Can we use the following naming convention (remove pinctrl, add gpio to match the datasheet convention) ?
> 
> ```
> gpio200_adc0_in0_default
> gpio201_adc0_in1_default
> ```

See my comment here https://github.com/zephyrproject-rtos/zephyr/pull/41888/files#r798373887

jvasanth1 replied:
Please review proposed naming in the [comment](https://github.com/zephyrproject-rtos/zephyr/pull/41888/commits/00fb2c6b656714a495e77d8e53db99f812925d64#r798974307).

tejlmand commented on soc/arm/microchip_mec/common/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_include_directories_ifdef(CONFIG_SOC_SERIES_MEC172X .)
+zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_MEC172X
+  soc_i2c.c
```
minor nit.

Line is not that long so you must wrap it.

gmarull replied:
@tejlmand maybe https://github.com/cheshirekow/cmake_format :-)?

tejlmand replied:
Thanks, worth taking a look at.

gmarull commented on dts/arm/microchip/mec172x/mec172xnsz-pinctrl.dtsi:
```diff
@@ -0,0 +1,794 @@
+				pinmux = < MCHP_XEC_PINMUX(0124, MCHP_AF4) >;
+			};
+
+			/* I2C port 0 */
+			pinctrl_i2c_scl0: i2c_scl0 {
+				pinmux = < MCHP_XEC_PINMUX(004, MCHP_AF1) >;
+			};
+
+			pinctrl_i2c_sda0: i2c_sda0 {
+				pinmux = < MCHP_XEC_PINMUX(003, MCHP_AF1) >;
+			};
+
+			/* I2C port 1 */
+			pinctrl_i2c_scl1: i2c_scl1 {
+				pinmux = < MCHP_XEC_PINMUX(0131, MCHP_AF1) >;
+			};
+
+			pinctrl_i2c_scl1_alt: i2c_scl1_alt {
+				pinmux = < MCHP_XEC_PINMUX(073, MCHP_AF2) >;
+			};
```
I think signal names should be aligned with datasheet:

![image](https://user-images.githubusercontent.com/25011557/152313477-5004f095-9d2d-4d0a-b48b-027e96e8a67e.png)

datasheet uses "i2c01_scl", not "i2c_scl1".

jvasanth1 replied:
Could you confirm, this naming convention is okay:

```
        /* I2C port 1 */
	i2c01_scl: i2c01_scl {
		pinmux = < MCHP_XEC_PINMUX(0131, AF1) >;
	};

       i2c01_scl_alt: i2c01_scl_alt {
		pinmux = < MCHP_XEC_PINMUX(073, AF2) >;
	}
```;

jvasanth1 replied:
Please confirm, so that we can proceed to name pins as in the datasheet. Also note, we are not appending `default`.

gmarull replied:
Looks good, but since you're not appending `default`, please, confirm the solution you expect to use for other states, such as `sleep`. It is important for nodes approach, since nodes can't be re-used for more than one state. It's a limitation people may not be aware since it is still not used by any driver, so it's better to think about. For example STM32 will use this approach:

```devicetree
/** autogenerated file **/

/* default state for i2c0 */
i2c0_scl_pa0 {
   pinmux = <STM32_PINMUX('A', 0, AF5);
};

i2c0_sda_pa1 {
    pinmux = <STM32_PINMUX('A', 1, AF5);
};

/* any pin is put into LP mode by using analog mode nodes,
 * this is how STM32 devices configure a pin for LP mode
 */
analog_pa0 {
    pinmux = <STM32_PINMUX('A', 0, ANALOG);
};

analog_pa1 {
    pinmux = <STM32_PINMUX('A', 1, ANALOG);
};

/** board file **/
&i2c0 {
    pinctrl-0 = <&i2c0_scl_pa0 &i2c0_sda_pa1>;
    pinctrl-0 = <&analog_pa0 &analog_pa1>;
    pinctrl-names = "default", "sleep";
};
```

jvasanth1 replied:
Thanks for the example.

Here is the updated naming convention we propose:


        ```
       /* I2C port 1 */
	i2c01_scl_gpio131: i2c01_scl_gpio131 {
		pinmux = < MCHP_XEC_PINMUX(0131, AF1) >;
	};

       i2c01_scl_alt_gpio073: i2c01_scl_alt_gpio073 {
		pinmux = < MCHP_XEC_PINMUX(073, AF2) >;
	};
```

For the later sleep mode support, we would add the following

```
	sleep_gpio131: sleep_gpio131 {
		pinmux = < MCHP_XEC_PINMUX(0131, AF0) >;
	};

       sleep_gpio073: sleep_gpio073 {
		pinmux = < MCHP_XEC_PINMUX(073, AF0) >;
	};
```

jvasanth1 replied:
Please confirm

gmarull replied:
Naming LGTM.

jvasanth1 replied:
Updated naming convention as per the discussion.

gmarull commented on dts/bindings/pinctrl/microchip,xec-pinctrl-conf.yaml:
```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2021 Microchip Technology Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
```
where is this file used now?

jvasanth1 replied:
This file is no longer used, will be removed.

gmarull commented on dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:
```diff
@@ -0,0 +1,116 @@
+    Microchip XEC Pin controller Node
+    Based on pincfg-node.yaml binding.
+    The MCHP XEC pin controller is a singleton node responsible for controlling
+    pin function selection and pin properties. For example, you can use this
+    node to select peripheral pin functions. For example, I2C
```
`For example, I2C`....? seems like it's missing content

jvasanth1 replied:
that is extraneous. will remove `For example, I2C`

gmarull commented on dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:
```diff
@@ -0,0 +1,116 @@
+      };
+
+    All device pin configurations should be placed in child nodes of the
+    'pinctrl' node, as shown in this example:
```
which example?

jvasanth1 replied:
will modify to `as shown in the spi0 example below`

gmarull commented on dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:
```diff
@@ -0,0 +1,116 @@
+        pinctrl-names = "default";
+     }
+
+     &pinctrl_qint_cs0 {
+        drive-open-drain;
+     };
```
mention that these should preferrably go into `-pinctrl.dtsi` files.

gmarull commented on dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:
```diff
@@ -0,0 +1,116 @@
+          type: string
+          default: "hw-default"
+          enum:
+          - "hw-default"        # Default value
```
```suggestion
          - "hw-default"
```

gmarull commented on dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:
```diff
@@ -0,0 +1,116 @@
+      pin in the board's DTS file and set the properties.
+
+      &spi0 {
+        pinctrl-0 = < &pinctrl_qint_cs0
+                      &pinctrl_qint_clk
+                      &pinctrl_qint_io0
+                      &pinctrl_qint_io1
+                      &pinctrl_qint_wpn >;
```
Please explain the naming convention:

![image](https://user-images.githubusercontent.com/25011557/152314401-ff6d618b-385a-4ed3-b602-a116806b7de8.png)

By looking at datasheet I could understand that `<signal>`, `<signal>_alt` and `<signal>_alt2` are expected node names (no need to specify the `gpio_xxx`), but I couldn't get that from the current binding docs. Since pinctrl DT part has many vendors-specific bits, documentation needs to be clear.

jvasanth1 replied:
This table is just listing those functions that are available on multiple pins to ease board layout.

Please see proposed changes in this [comment ](https://github.com/zephyrproject-rtos/zephyr/pull/41888#discussion_r798974307)and confirm.

jvasanth1 replied:
Naming convention updated as per review comments.

albertofloyd commented on drivers/spi/spi_xec_qmspi_ldma.c:
```diff
@@ -1242,14 +1095,14 @@ static int qmspi_xec_init(const struct device *dev)
 #endif
 
-	int ret = xec_qspi_config_port();
+	z_mchp_xec_pcr_periph_sleep(cfg->pcr_idx, cfg->pcr_pos, 0);
```
Does 17xx soc header have MCHP_PCR_SLEEP_DIS macros?

jvasanth1 replied:
172x soc headers do not have these macros. The code reviewers (at the time the 172x soc was checked in) demanded no macros for register access. They wanted the functionality in drivers.

gmarull commented on soc/arm/microchip_mec/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,114 @@
+
+/** Type for MCHP XEC pin. */
+typedef struct pinctrl_soc_pin {
+	/** Pinmux settings (port, pin and function). */
+	uint32_t pinmux;
+	/** Pin configuration (bias, drive and slew rate). */
+	uint32_t pincfg;
```
This is really an opaque type, so can be modified at any time with no application/driver effects, except pinctrl driver, of course. So, I'd suggest not to waste unnecessary ROM space. If some future families need more ROM space, you could even have another opaque time.

jvasanth1 replied:
Updated to use `uint16_t`

albertofloyd commented on boards/arm/mec172xevb_assy6906/mec172xevb_assy6906.dts:
```diff
@@ -76,14 +77,34 @@
+
+/* If ESPI_CS is open drain */
+&pinctrl_espi_cs {
+	drive-open-drain;
```
@jvasanth1 I recall from MEC152x that this is not required. Is this different in MEC172x? 

This since the eSPI Host can select either push-pull or open-drain for this pin during eSPI bus enumeration the MCHP ESPI HW block must be able to dynamically change the pin’s output buffer type.  
This is not reflected in the GPIO control register.  This is one of the very few pins that has a pin feature automatically configured.

jvasanth1 replied:
Yes, this is not required for the two pins (ESPI_CS and ESPI_ALERT), will remove from the board .dts

albertofloyd commented on boards/arm/mec172xevb_assy6906/mec172xevb_assy6906.dts:
```diff
@@ -76,14 +77,34 @@
+
+/* If ESPI_ALERT is open drain */
+&pinctrl_espi_alert {
+	drive-open-drain;
```
Same than above

jvasanth1 replied:
will remove this

albertofloyd commented on boards/arm/mec172xevb_assy6906/mec172xevb_assy6906.dts:
```diff
@@ -130,21 +154,68 @@
+	pinctrl-names = "default";
+};
+
+&pinctrl_qshd_cs0 {
```
@jvasanth1 per my debug this seems to cause the glitch in GPIO055 (RSMRST) in MAF mode. 
This was not part of MEC152x pinmux. https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/mec15xxevb_assy6853/pinmux.c#L428

Could you clarify this addition?

jvasanth1 replied:
This was added to configure the pins for shared spi application to test on the EVB board; 
but we see the problem now in MAF mode since this pin is used as RSMRST. 

I will remove this and then application can configure it in their .dts file if required.

jvasanth1 commented on drivers/pinctrl/pinctrl_mchp_xec.c:
```diff
@@ -0,0 +1,83 @@
+		return -ENODEV;
+	}
+
+	return gpio_mchp_xec_configure(port_device, pin, cfg, altf);
```
fixed

mbolivar-nordic commented on dts/arm/microchip/mec172xnsz.dtsi:
```diff
@@ -287,71 +287,78 @@
+			#size-cells = <1>;
+			reg = <0x40081000 0x1000>;
+
+			gpio_000_036: gpio@40081000 {
```
Why are the GPIO controllers being moved into the pin control node? I am surprised by this. The two subsystems are relatively independent within zephyr and I'm not aware of any other devicetrees that behave this way. Please justify this.

jvasanth1 replied:
The gpio_000_xxx nodes are accessed by the pinctrl driver to configure the pin. Earlier we used the node handles directly, but it was suggested to use phandle instead in this [comment](https://github.com/zephyrproject-rtos/zephyr/pull/41888#discussion_r791047583).
Hence we updated to use phandle and moved the gpio_000_xxx nodes.

gmarull replied:
The same happens for STM32/GD32 (inherited from Linux, see https://elixir.bootlin.com/linux/latest/source/arch/arm/boot/dts/stm32mp131.dtsi#L167). In such case, there's no pinctrl device, really. It's all done via the GPIO registers.

mbolivar-nordic replied:
OK, thanks for the justification. This is resolved.

jvasanth1 replied:
As per latest feedback from @gmarull we don't need the gpio_xxx_xxx nodes inside the pinctrl node. We are going to our original approach where we don't use phandle.
Hence we will be moving the gpio_xxx_xxx nodes outside the pinctrl node.

mbolivar-nordic commented on dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:
```diff
@@ -0,0 +1,126 @@
+        slew-rate:
+          required: false
+          type: string
+          default: "low-speed"
```
All use of defaults must be justified. Please read the rules in https://docs.zephyrproject.org/latest/guides/dts/bindings.html#rules-for-mainline-bindings and update this PR accordingly

jvasanth1 replied:
Will add default value in the description

jvasanth1 replied:
Updated description with default value.

mbolivar-nordic replied:
Sorry, but that wasn't my point. I am not asking for the default to be placed into the description. I am asking for the default to be _justified_, according to the rules in my above link. Please revisit.

jvasanth1 replied:
The pinctrl driver will not program the slew-rate (or the drive-strength) if the those properties are not defined in the dts. As such we are considering to remove the default field (for slew-rate and drive-strength) in the yaml file.
Can you confirm?

Note: we don't want to program the slew-rate (or drive-strength) for each pin used; we want to use the hardware defaults most of the time.

gmarull replied:
A question here: if "hardware defaults" mean no-op, how do I know that pin is correctly configured? What if e.g. a bootloader configures pins with a different set of options?

mbolivar-nordic replied:
I have the same question, but I want to clarify something else.

It's OK to use `default` in the binding, if and only if you also justify **why** the default was chosen in the description, **and** that justification follows the rules in my link. So keeping the default is OK -- especially to cover cases like the bootloader situation in @gmarull 's question -- you just have to justify it.

Just writing "the default value of X is the SoC power-on-reset value" is fine as long as the value is independently configurable from other values  as described in the bindings documentation.

Is that clear?

jvasanth1 replied:
Okay, we will keep default in the binding and add "The default value of slew-rate is the SoC power-on-reset value" in the description.
Yes, this value is independently configurable from other values.

jvasanth1 replied:
> A question here: if "hardware defaults" mean no-op, how do I know that pin is correctly configured? What if e.g. a bootloader configures pins with a different set of options?

The pins will be in the hardware default state (as in the datasheet) except for those configured by the boot rom loader. We are working on how to make available this information (boot rom configured pins). 

Can we share the boot rom configured pins table in [CPGZephyrDocs/MEC172x]([CPGZephyrDocs](https://github.com/MicrochipTech/CPGZephyrDocs)/MEC172x/) and update the board's [rst](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/mec172xevb_assy6906/doc/index.rst) file with its link?
If so, I will be creating a separate PR for updating the .rst.

mbolivar-nordic replied:
> Okay, we will keep default in the binding and add "The default value of slew-rate is the SoC power-on-reset value" in the description.
> Yes, this value is independently configurable from other values.

Sounds good, thanks. I see you're making a lot of contributions lately, which is great! Please remember the rules for upstream bindings in my link and apply them consistently.

gmarull commented on drivers/gpio/gpio_mchp_xec_v2.c:
```diff
@@ -368,6 +378,128 @@ static void gpio_gpio_xec_port_isr(const struct device *dev)
+	default:
+		break;
+	}
+
+	if ((conf >> MCHP_XEC_OTYPER_POS) & MCHP_XEC_OTYPER_MASK) {
+		flags |= GPIO_LINE_OPEN_DRAIN;
+	}
+
+	temp = (conf >> MCHP_XEC_OVAL_POS) & MCHP_XEC_OVAL_MASK;
+	if (temp) {
+		flags |= GPIO_OUTPUT;
+		if (temp == MCHP_XEC_OVAL_DRV_HIGH) {
+			flags |= GPIO_OUTPUT_INIT_HIGH;
+		} else {
+			flags |= GPIO_OUTPUT_INIT_LOW;
+		}
+	} else {
+		flags |= GPIO_INPUT;
+	}
+
+	if (conf & BIT(MCHP_XEC_PIN_LOW_POWER_POS)) {
+		flags = GPIO_DISCONNECTED;
+	}
+
+	ret = gpio_xec_configure(dev, pin, flags);
+	if (ret) {
+		return ret;
+	}
+
+	if (flags != GPIO_DISCONNECTED) {
+		/* program slew rate and drive strength */
+		config_drive_slew(dev, pin, conf);
+
+		/* program alternate function */
+		config_alt_func(dev, pin, altf);
+	}
+
+	return 0;
+}
+
```
This should all be moved to the pinctrl driver (which is where this is used, really). As an example, having code that configures alternate function in a GPIO driver doesn't make much sense. And this also creates a dependency between pinctrl & GPIO for no reason. STM32 does something similar, but mostly for historical reasons, and should probably be fixed. GD32 platform (similar/cloned hardware) has independent GPIO/pinctrl drivers.

jvasanth1 replied:
@gmarull 
Yes will move the three functions marked to pinctrl driver.

Could you review other all other commits in this PR and provide your comments? We like to get this PR approved by you in the next iteration of commits.

jvasanth1 replied:
Moved these functions to pinctrl driver.

gmarull commented on dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:
```diff
@@ -0,0 +1,126 @@
+            required: true
+            description: Pinmux selection
+
+        gpioph:
+            type: phandle
+            required: true
+            description: phandle to GPIO node
```
where is this used?

jvasanth1 replied:
This is used in pinctrl_mchp_xec.c to retrieve the gpio_xxx_xxx node associated with each pin.

gmarull commented on dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:
```diff
@@ -0,0 +1,126 @@
+    - slew-rate
+
+    Driver strength and slew rate hardware defaults vary by SoC and pin.
+    The "hw-default" setting tell the pinctrl code to not touch the slew
+    rate or drive strength settings for the respective pin.
```
`hw-default` is not present anymore, this should be removed then?

jvasanth1 replied:
yes, will remove

jvasanth1 replied:
removed these two lines

gmarull commented on drivers/pinctrl/Kconfig.xec:
```diff
@@ -0,0 +1,9 @@
+config PINCTRL_MCHP_XEC
+	bool "Pin controller driver for MCHP XEC MCUs"
+	depends on SOC_SERIES_MEC172X
+	default y
```
please default based on `dt_compat_status` (see other drivers as an example)

jvasanth1 replied:
updated

gmarull commented on drivers/pinctrl/pinctrl_mchp_xec.c:
```diff
@@ -0,0 +1,63 @@
+		cfg = pins[i].pincfg;
+		portpin = MEC_XEC_PINMUX_PORT_PIN(mux);
+
+		dev = (const struct device *)pins[i].gpioph;
```
I feel this is mixing GPIO/pinctrl stuff really. Other platforms such as STM32/GD32 where pinmux is done using GPIO registers don't do anything like this. I'd prefer to stick with the initial approach.

jvasanth1 replied:
The phandle was introduced to address Maureen's [comment](https://github.com/zephyrproject-rtos/zephyr/pull/41888#discussion_r791047583). 

Can you confirm you are okay with not using phandle and using the `gpio_ports[]` table as before?

jvasanth1 replied:
Another approach (as suggested in the [comment](https://github.com/zephyrproject-rtos/zephyr/pull/41888#discussion_r791047583)) is to resolve this issue  as a separate PR. 

We would like to proceed with the phandle approach for now.

jvasanth1 replied:
Thanks for confirming; will revert to previous solution as described in the [comment](https://github.com/zephyrproject-rtos/zephyr/pull/41888#discussion_r809409996).

gmarull commented on drivers/pinctrl/pinctrl_mchp_xec.c:
```diff
@@ -0,0 +1,63 @@
+{
+	uint32_t pin = MCHP_XEC_PINMUX_PIN(portpin);
+
+	if (!dev || !device_is_ready(dev)) {
+		return -ENODEV;
+	}
```
related to previous comment: this checks for device readyness every time state is configured, doesn't make much sense.

jvasanth1 replied:
will remove device_is_ready() check

jvasanth1 replied:
updated

gmarull commented on include/dt-bindings/pinctrl/mchp-xec-pinctrl.h:
```diff
@@ -0,0 +1,81 @@
+ * b[11:9] = function
+ */
+#define MCHP_XEC_PINMUX(n, f)						      \
+	(((((n)>>5) & MCHP_XEC_PINMUX_PORT_MSK) << MCHP_XEC_PINMUX_PORT_POS) |\
```
nit ` >> 5` (spaces)

jvasanth1 replied:
Fixed

gmarull commented on dts/arm/microchip/mec172x/mec172xnsz-pinctrl.dtsi:
```diff
@@ -0,0 +1,1311 @@
+&pinctrl {
+	/* ADC */
+	adc00_gpio200: adc00_gpio200 {
+		pinmux = < MCHP_XEC_PINMUX(0200, MCHP_AF1) >;
+		gpioph = < &gpio_200_236 >;
```
This is both mixing GPIO and pinctrl, and, it duplicates information: 0200. Please just go with previous approach as it made more sense, IMHO.

jvasanth1 replied:
Previous approach involved using a gpio_ports[] table in the [pinctrl_mchp_xec.c](https://github.com/MicrochipTech/zephyr/blob/mec172x_pinctrl_mcreview1/drivers/pinctrl/pinctrl_mchp_xec.c)

Can you confirm we can revert back to that approach? 
This way the gpioph will be removed in the dtsi file as well; and there is no need for the gpio_xxx_xxx nodes within the pinctrl node as mentioned in the [comment](https://github.com/zephyrproject-rtos/zephyr/pull/41888#discussion_r808436043).

Please confirm.

jvasanth1 replied:
Removed phandle and reverted to original approach of using gpio_ports[].

jvasanth1 commented:
Resolved all review comments in latest push; please review and approve.

mbolivar-nordic commented on dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:
```diff
@@ -0,0 +1,120 @@
+      };
+
+    All device pin configurations should be placed in child nodes of the
+    'pinctrl' node, as shown in this example:
```
@gmarull's comment is still open; this should not have been resolved

jvasanth1 replied:
will fix

jvasanth1 replied:
fixed in PR #43282

mbolivar-nordic commented on dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:
```diff
@@ -0,0 +1,120 @@
+    Driver strength and slew rate hardware defaults vary by SoC and pin.
+
+    An example for MEC172x family, include the chip level pinctrl
+    DTSI file in the board level DTS:
+
+      #include <microchip/mec172x/mec172xnsz-pinctrl.dtsi>
+
+      We want to use the shared SPI port of the MEC172x QMSPI controller
+      and want the chip select 0 to be open-drain.
+
+      To change a pin's pinctrl default properties add a reference to the
+      pin in the board's DTS file and set the properties.
```
The indentation is off here.

jvasanth1 replied:
I think line 43 is fine; will fix line 45 onwards.

mbolivar-nordic commented on dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:
```diff
@@ -0,0 +1,120 @@
+child-binding:
+    description: |
+      This binding gives a base representation of the Microchip XEC pins
+      configration
```
configuration

jvasanth1 commented:
Rebased and pushed latest changes to [mec172x_pinctrl](https://github.com/MicrochipTech/zephyr/tree/mec172x_pinctrl) branch; however don't see this PR getting updated.

jvasanth1 commented:
should I submit as a new PR?

gmarull commented:
> should I submit as a new PR?

this should not be needed, just rebase against `upstream/main` and force-push.

jvasanth1 commented:
Tried it, didn't work. The updated branch is [here](https://github.com/MicrochipTech/zephyr/commits/mec172x_pinctrl).

Also after the recent rebase, I am getting my compilation aborted due to Kconfig warnings:
`warning: the default selection LV_COLOR_DEPTH_1 (undefined) of <choice_LV_COLOR_DEPTH_BITS> (defined at ....) is not contained in the choice.`

gmarull commented:
> Tried it, didn't work. The updated branch is [here](https://github.com/MicrochipTech/zephyr/commits/mec172x_pinctrl).
> 
> Also after the recent rebase, I am getting my compilation aborted due to Kconfig warnings: `warning: the default selection LV_COLOR_DEPTH_1 (undefined) of <choice_LV_COLOR_DEPTH_BITS> (defined at ....) is not contained in the choice.`

it's likely because of https://github.com/zephyrproject-rtos/zephyr/issues/43163, just run `west update` in-tree.

jvasanth1 commented:
`west update` and rebase resolved the compilation error and I forced pushed, however I still don't see the PR getting updated.

jvasanth1 commented:
shall I submit as a new PR?

mbolivar-nordic commented:
> it's likely because of #43163, just run `west update` in-tree.

The fix for this is now merged; I suggest rebasing on top.

jvasanth1 commented:
rebased and pushed again; still don't see the PR getting updated.

jvasanth1 commented:
I would like to open a new PR; Please confirm

gmarull commented:
> I would like to open a new PR; Please confirm. Is the push not taking effect due to the conflict message below? I have rebased with main and fixed the merge conflict due to deletion of the pinmux.c file.

Never seen that case before, may be related with this rebase accident:

![image](https://user-images.githubusercontent.com/25011557/155765937-6ff0e990-dd0f-4fcd-853c-aadd9ef79e61.png)

It's a shame we lose track of current review comments because of a new PR. I guess it's ok if there's no other way. @mbolivar-nordic any ideas on why this could happen?

jvasanth1 commented:
Created new PR #43282 

Please review and approve.

gmarull commented:
Closing, see https://github.com/zephyrproject-rtos/zephyr/pull/43282 instead.
